add explicit vectorization for Half dtype on CPU

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96082
* #96081
* #96080
* #96079
* #96078
* #96077
* __->__ #96076

This patch is part of half float performance optimization on CPU:
* add specification for dtype `Half` in `Vectorized<>` under both avx256 and avx512.
* add specification for dtype `Half` in functional utils, e.g. `vec::map_reduce<>()`, which uses float32 as accumulate type.

Also add a helper struct `vec_hold_type<scalar_t>`, since Vectorized<Half>::value_type is pointing to its underlying storage type which is `uint16_t`, leading to error if the kernel uses `Vec::value_type`.

Half uses the same logic as BFloat16 in the Vectorized<>, each half vector is mapped to 2x float vectors for computation.

Notice that this patch modified the cmake files by adding **-mf16c** on AVX2 build, from https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html, we can see that all the hardware platforms that support **avx2** already have **f16c**

cc @jgong5 @XiaobingSuper @sanchitintel @ashokei @jingxu10

mingfeima committed:
add explicit vectorization for Half dtype on CPU

[ghstack-poisoned]

mingfeima committed:
Update on "add explicit vectorization for Half dtype on CPU"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add explicit vectorization for Half dtype on CPU"




cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add explicit vectorization for Half dtype on CPU"


This patch is part of half float performance optimization on CPU:
* add specification for dtype `Half` in `Vectorized<>` under both avx256 and avx512.
* add specification for dtype `Half` in functional utils, e.g. `vec::map_reduce<>()`, which uses float32 as accumulate type.

Also add a helper struct `vec_hold_type<scalar_t>`, since Vectorized<Half>::value_type is pointing to its underlying storage type which is `uint16_t`, leading to error if the kernel uses `Vec::value_type`.

Half uses the same logic as BFloat16 in the Vectorized<>, each half vector is mapped to 2x float vectors for computation.

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

malfet commented on aten/src/ATen/cpu/vec/vec512/vec512_float16.h:
```diff
@@ -0,0 +1,887 @@
+    }
+  }
+  template <int64_t mask>
+  static Vectorized<Half> blend(const Vectorized<Half>& a, const Vectorized<Half>& b) {
```
This looks like exact duplicate of https://github.com/pytorch/pytorch/blob/191a2322f01cf8646a916bdbdcdbfb352a9cc263/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h#L122-L126

Can it be avoided somehow?

jgong5 replied:
Not only this function, but for others, can we share the implementation between bf16 and fp16 somehow, e.g., create a common base class. There are three kinds of functions:
1. Share exact same implementation, which can be put into the base.
2. Share the main fp32 compute but with conversion ops in the prefix/suffix. The main compute body can be put into the base.
3. Different implementation, put as overrides in the subclasses.

mingfeima replied:
I have abstracted the common code to `struct Vectorized16` to reduce the duplicate code between `Vectorized<BFloat16>` and `Vectorized<Half>`. And used some pre processor on some occasions.

malfet commented on aten.bzl:
```diff
@@ -3,7 +3,7 @@ load("@rules_cc//cc:defs.bzl", "cc_library")
 CPU_CAPABILITY_NAMES = ["DEFAULT", "AVX2"]
 CAPABILITY_COMPILER_FLAGS = {
-    "AVX2": ["-mavx2", "-mfma"],
+    "AVX2": ["-mavx2", "-mfma", "-mf16c"],
```
Can you please add some reference to the doc that `-mfc16c` is always available whenever `-mavx2` is is available?

mingfeima replied:
Oh this is really important, i should add more comment. I have parsed the gcc compilation flag with **-march** from https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html

**FP16C** is added after **AVX**, but before **AVX2**. If the hardware has **AVX2**, it has **FP16C**. I have listed the detailed info in a table below.

malfet commented on aten/src/ATen/cpu/vec/functional_base.h:
```diff
@@ -83,7 +83,8 @@ inline scalar_t vec_reduce_all(const Op& vec_fun, const Vectorized<scalar_t>& ac
 
-template <typename scalar_t, typename Op>
+template <typename scalar_t, typename Op,
+          typename std::enable_if<!is_reduced_floating_point<scalar_t>::value, int>::type = 0>
```
In typical C++17 fashion, can you please add `is_reduced_floating_point_v` constexpr to avoid typing ::value. 
Also, use `enable_if_t` which is available since C++14
```suggestion
          typename std::enable_if_t<!is_reduced_floating_point<scalar_t>::value, int> = 0>
```

mingfeima replied:
Thanks for the hint! Updated!

malfet commented on aten/src/ATen/cpu/vec/vec256/vec256_float16.h:
```diff
@@ -0,0 +1,785 @@
+    }
+  }
+  template <int64_t mask>
+  static Vectorized<Half> blend(const Vectorized<Half>& a, const Vectorized<Half>& b) {
```
That's the 2nd time I encounter the same blend function. Can it be generalized somehow?

mingfeima replied:
duplicate code removed!

malfet reviewed (approved):
Overall looks fine, but please reply on how code duplication can be avoided. If this can't be templatized, perhaps using preprocessor will do.
Also, use C++14/C++17 language features when possible

mingfeima committed:
Update on "add explicit vectorization for Half dtype on CPU"


This patch is part of half float performance optimization on CPU:
* add specification for dtype `Half` in `Vectorized<>` under both avx256 and avx512.
* add specification for dtype `Half` in functional utils, e.g. `vec::map_reduce<>()`, which uses float32 as accumulate type.

Also add a helper struct `vec_hold_type<scalar_t>`, since Vectorized<Half>::value_type is pointing to its underlying storage type which is `uint16_t`, leading to error if the kernel uses `Vec::value_type`.

Half uses the same logic as BFloat16 in the Vectorized<>, each half vector is mapped to 2x float vectors for computation.

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima commented:
@malfet @jgong5 since this patch modifies the cmake file (add **f16c** to AVX2 build), add some comments here: all the hardware platforms that has **avx2** has **f16c** support.

source info: https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html

platform | vendor | AVX | FP16C | AVX2 | AVX512
-- | -- | -- | -- | -- | --
ivybridge | intel | Yes | Yes | No | No
haswell | intel | Yes | Yes | Yes | No
broadwell | intel | Yes | Yes | Yes | No
skylake | intel | Yes | Yes | Yes | Yes
cannonlake | intel | Yes | Yes | Yes | Yes
icelake | intel | Yes | Yes | Yes | Yes
cascadelake | intel | Yes | Yes | Yes | Yes
cooperlake | intel | Yes | Yes | Yes | Yes
tigerlake | intel | Yes | Yes | Yes | Yes
sapphirerapids | intel | Yes | Yes | Yes | Yes
alderlake | intel | Yes | Yes | Yes | No
rocketlake | intel | Yes | Yes | Yes | Yes
graniterapids | intel | Yes | Yes | Yes | Yes
bdver2,3,4 | amd | Yes | Yes | No | No
znver1,2,3 | amd | Yes | Yes | Yes | No
znver4 | amd | Yes | Yes | Yes | Yes
btver2 | amd | Yes | Yes | No | No

mingfeima committed:
Update on "add explicit vectorization for Half dtype on CPU"


This patch is part of half float performance optimization on CPU:
* add specification for dtype `Half` in `Vectorized<>` under both avx256 and avx512.
* add specification for dtype `Half` in functional utils, e.g. `vec::map_reduce<>()`, which uses float32 as accumulate type.

Also add a helper struct `vec_hold_type<scalar_t>`, since Vectorized<Half>::value_type is pointing to its underlying storage type which is `uint16_t`, leading to error if the kernel uses `Vec::value_type`.

Half uses the same logic as BFloat16 in the Vectorized<>, each half vector is mapped to 2x float vectors for computation.

Notice that this patch modified the cmake files by adding **-mf16c** on AVX2 build, from https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html, we can see that all the hardware platforms that support **avx2** already have **f16c**

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima commented on aten/src/ATen/cpu/vec/vec512/vec512_bfloat16.h:
```diff
@@ -854,75 +915,345 @@ Vectorized<BFloat16> inline fmadd(const Vectorized<BFloat16>& a,
+Vectorized<Half> inline Vectorized<Half>::operator!=(const Vectorized<Half>& other) const {
+  return binary_op_as_fp32(*this, other, [](__m512 x, __m512 y) {
+    auto zero_vec = _mm512_set1_epi32(0);
+    auto cmp = _mm512_cmp_ps_mask(x, y, _CMP_NEQ_OQ);
```
thanks!

mingfeima commented on aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:
```diff
@@ -564,16 +624,16 @@ Vectorized<BFloat16> inline Vectorized<BFloat16>::operator!=(const Vectorized<BF
 
 Vectorized<BFloat16> inline operator+(const Vectorized<BFloat16>& a, const Vectorized<BFloat16>& b) {
-  return bfloat16_binary_op_as_fp32(a, b, [](const __m256& x, const __m256& y) { return _mm256_add_ps(x, y); });
+  return binary_op_as_fp32(a, b, [](const __m256& x, const __m256& y) { return _mm256_add_ps(x, y); });
```
this one is not a member function of `Vectorized<BFloat16>`, can't fold it into the base type.

mingfeima committed:
Update on "add explicit vectorization for Half dtype on CPU"


This patch is part of half float performance optimization on CPU:
* add specification for dtype `Half` in `Vectorized<>` under both avx256 and avx512.
* add specification for dtype `Half` in functional utils, e.g. `vec::map_reduce<>()`, which uses float32 as accumulate type.

Also add a helper struct `vec_hold_type<scalar_t>`, since Vectorized<Half>::value_type is pointing to its underlying storage type which is `uint16_t`, leading to error if the kernel uses `Vec::value_type`.

Half uses the same logic as BFloat16 in the Vectorized<>, each half vector is mapped to 2x float vectors for computation.

Notice that this patch modified the cmake files by adding **-mf16c** on AVX2 build, from https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html, we can see that all the hardware platforms that support **avx2** already have **f16c**

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add explicit vectorization for Half dtype on CPU"


This patch is part of half float performance optimization on CPU:
* add specification for dtype `Half` in `Vectorized<>` under both avx256 and avx512.
* add specification for dtype `Half` in functional utils, e.g. `vec::map_reduce<>()`, which uses float32 as accumulate type.

Also add a helper struct `vec_hold_type<scalar_t>`, since Vectorized<Half>::value_type is pointing to its underlying storage type which is `uint16_t`, leading to error if the kernel uses `Vec::value_type`.

Half uses the same logic as BFloat16 in the Vectorized<>, each half vector is mapped to 2x float vectors for computation.

Notice that this patch modified the cmake files by adding **-mf16c** on AVX2 build, from https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html, we can see that all the hardware platforms that support **avx2** already have **f16c**

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima commented:
@pytorchbot merge

tbennun commented:
@mingfeima Merging this PR breaks IBM POWER builds.
